src/Visitors/AbstractVisitor.php

```diff
@@ -32,48 +49,32 @@ public function leaveNode(Node $node): Node
     {
         if ($this->shouldInsertNode($node) && !$this->isNodeExists) {
             $this->insertNode($node);
+        }
 
-            if ($node instanceof Class_) {
-                return $this->rebuildClass($node);
-            }
+        if ($this->isNodeExists && $this->shouldUpdateNode($node)) {
+            $this->updateNode($node);
         }
 
         return $node;
     }
 
-    /**
-     * Used to automatically group statements by their type, which simplifies development 
-     * because we don't need to define the correct place for the insertable node.
-     */
-    protected function rebuildClass(Class_ $node): Class_
+    protected function getInsertIndex(array $stmts, string $insertType): int
     {
-        $factory = new BuilderFactory();
-        $classBuilder = $factory->class($node->name->toString());
-
-        if ($node->getDocComment()) {
-            $classBuilder->setDocComment($node->getDocComment());
-        }
-
-        if ($node->extends) {
-            $classBuilder->extend($node->extends);
-        }
-
-        if (!empty($node->implements)) {
-            $classBuilder->implement(...$node->implements);
-        }
-
-        if ($node->isAbstract()) {
-            $classBuilder->makeAbstract();
-        }
-
-        if ($node->isFinal()) {
-            $classBuilder->makeFinal();
-        }
-
-        foreach ($node->stmts as $stmt) {
-            $classBuilder->addStmt($stmt);
+        $insertIndex = 0;
+        $insertTypeOrder = array_search($insertType, self::TYPE_ORDER);
+
+        foreach ($stmts as $i => $stmt) {
+            foreach (self::TYPE_ORDER as $type) {
+                if ($stmt instanceof $type) {
+                    $currentTypeOrder = array_search($type, self::TYPE_ORDER);
```

Comment:
will it working the same? If yes - let's name it `$currentTypeIndex` instead of `$currentTypeOrder`

```suggestion
            foreach (self::TYPE_ORDER as $currentTypeOrder => $type) {
                if ($stmt instanceof $type) {
```